### PR TITLE
Bug 1957349: Avoid machines going into failed phase if unexpected ProvisioningState appears 

### PR DESCRIPTION
**Pull request by lobziik**
In case if azure vm provisioning state different from "Updating", "Creating" or "Succeded" machine treat as non-existent after existence check. 
In combo with this https://github.com/kubernetes/kubectl/issues/564 issue it appears to be quite nasty...

This patch introduces less strict treatment of ProvisioningState vm's field.

### CODE REVIEW

**Comment by JoelSpeed** [comment]
Is there an equivalent piece of code in the upstream provider?

I'm wondering if instead of enumerating all of the states and saying these are ok, we should just enumerate the ones we know to be bad and reverse the logic here, WDYT? When there's an explicit list of ok values, we are more likely to have bugs than a known list of bad values IMO

**Comment by lobziik** [comment]
@JoelSpeed not really. https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/azure/services/virtualmachines/virtualmachines.go#L90 here is CAPZ part which handles virtual machines. 

Basically they ProvisioningState just goes to machine State here: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/azure/converters/vm.go#L30

Agree with your proposal. Will update this pr.

File: pkg/cloud/azure/actuators/machine/reconciler.go
Comment (JoelSpeed) [inline pkg/cloud/azure/actuators/machine/reconciler.go:393]: Looking at the list, what about deallocating/deallocate? Does this need to be accounted for? I think this is what happens when the VM is being deleted outside of the scope of Machine API, in which case I think we want to go failed in this case, so I think these should also be included
```suggestion
	switch vmProvisioningState {
	case v1beta1.VMStateFailed, v1beta1.VMStateDeallocating, v1beta1.VMStateDeallocated:
```

Could use a switch and then interpolate the actual state into the log message?

https://github.com/openshift/cluster-api-provider-azure/blob/b024905873bbc955a432559a9af923a6f002a4ae/pkg/apis/azureprovider/v1beta1/types.go#L342-L345

**Review by JoelSpeed — COMMENTED** [review]
Would like to know how you've tested this manually before we merge, have you run this on a cluster and tried starting/stoppping/modifying/deleting the VM from the console to make sure that the controller acts appropriately?

**Comment by lobziik** [comment]
>Would like to know how you've tested this manually before we merge, have you run this on a cluster and tried starting/stoppping/modifying/deleting the VM from the console to make sure that the controller acts appropriately?

File: pkg/cloud/azure/actuators/machine/reconciler.go
Comment (lobziik) [reply pkg/cloud/azure/actuators/machine/reconciler.go:393]: I don't think that we need to account this here. This list is a mix between ProvisioningState and PowerState of vm. In case of vm deletion ProvisioningState would indicate `Deleting` for some period of time.

However `Deallocating/Deallocated` happens when vm was just powered off but not deleted. Not sure that we should treat it as non existing.

Here is a bit more context: https://github.com/openshift/cluster-api-provider-azure/pull/219/files#diff-af805e1e45f03df0b5b56ff4413e5ad52cd31904a94d37e8e916751953e4687dR295

**Comment by lobziik** [comment]
/hold
Went through old docs in internet archive, want to check one more thing before merging it.

**Comment by lobziik** [comment]
/hold cancel

https://web.archive.org/web/20201111230753/https://docs.microsoft.com/en-us/azure/virtual-machines/states-lifecycle - judging by this old documentation and my observations how ProvisioningState changes (for example - vm goes to provisioningState "Updating" when you attaching disk or network interface) I removed relying on this field at all and align logic with upstream one (e.g threat machine as non-existent if resource not found at all).

**Review by JoelSpeed — COMMENTED** [review]
/lgtm

I'd like to see some QE acknowledgement before we merge though as this is quite a large change with what we were doing before and we are very close to code freeze, I would hate to delay the release because this breaks something

**Review by Danil-Grigorev — APPROVED** [review]
/approve

**Comment by elmiko** [comment]
just wanted to note for the record

> I'd like to see some QE acknowledgement before we merge though as this is quite a large change with what we were doing before and we are very close to code freeze, I would hate to delay the release because this breaks something

@lobziik confirmed with @miyadav that this patch works

**Review by elmiko — COMMENTED** [review]
this makes sense, i have a small nit but perhaps it's just my confusion about the test.

since this has both labels, i'm gonna put a hold to give @lobziik a chance to respond. feel free to clear the hold if you add a response, or my question doesn't seem relevant.
/hold

File: pkg/cloud/azure/actuators/machine/reconciler_test.go
Comment (elmiko) [reply pkg/cloud/azure/actuators/machine/reconciler_test.go:60]: thinking about this a little more, maybe just change the name to `Delete while creating provisioning sate` ?

**Comment by elmiko** [comment]
oops, just realized it actually doesn't have both labels. i'll remove the hold
/hold cancel

File: pkg/cloud/azure/actuators/machine/reconciler_test.go
Comment (lobziik) [reply pkg/cloud/azure/actuators/machine/reconciler_test.go:60]: Oh, I just missed it... Will fix in couple minutes. Thanks for catching it!

**Review by elmiko — APPROVED** [review]
changes look really nice @lobziik , thanks!
/lgtm

**Pull request merged**
